File: crypto_agent/indicators.py
def compute_sma(closes: list[float], period: int) -> float:
    if len(closes) < period:
        return 0.0
    return sum(closes[-period:]) / period


def compute_rsi(closes: list[float], period: int = 14) -> float:
    if len(closes) < period + 1:
        return 50.0
    deltas = [closes[i] - closes[i - 1] for i in range(1, len(closes))]
    gains = [d if d > 0 else 0 for d in deltas[-period:]]
    losses = [-d if d < 0 else 0 for d in deltas[-period:]]
    avg_gain = sum(gains) / period
    avg_loss = sum(losses) / period
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def compute_bollinger(closes: list[float], period: int = 20, std_dev: float = 2.0) -> tuple[float, float, float]:
    if len(closes) < period:
        return 0.0, 0.0, 0.0
    recent = closes[-period:]
    mid = sum(recent) / period
    variance = sum((x - mid) ** 2 for x in recent) / period
    std = variance ** 0.5
    return mid - std_dev * std, mid, mid + std_dev * std


def generate_signals(closes: list[float], last_price: float) -> list[str]:
    signals = []
    sma20 = compute_sma(closes, 20)
    sma50 = compute_sma(closes, 50)
    rsi = compute_rsi(closes)
    bb_lower, bb_mid, bb_upper = compute_bollinger(closes)

    if sma20 and sma50:
        if sma20 > sma50:
            signals.append("MA Crossover: SMA20 > SMA50 → BULLISH")
        else:
            signals.append("MA Crossover: SMA20 < SMA50 → BEARISH")

    if rsi < 30:
        signals.append(f"RSI({rsi:.1f}): Oversold → BUY signal")
    elif rsi > 70:
        signals.append(f"RSI({rsi:.1f}): Overbought → SELL signal")
    else:
        signals.append(f"RSI({rsi:.1f}): Neutral")

    if bb_lower and bb_upper:
        if last_price < bb_lower:
            signals.append(f"Bollinger: Price below lower band ({bb_lower:.2f}) → BUY signal")
        elif last_price > bb_upper:
            signals.append(f"Bollinger: Price above upper band ({bb_upper:.2f}) → SELL signal")
        else:
            signals.append(f"Bollinger: Price within bands ({bb_lower:.2f} - {bb_upper:.2f})")

    return signals

File: src/indicators.ts
export function computeSma(closes: number[], period: number): number {
  if (closes.length < period) return 0;
  const slice = closes.slice(-period);
  return slice.reduce((a, b) => a + b, 0) / period;
}

export function computeRsi(closes: number[], period = 14): number {
  if (closes.length < period + 1) return 50;
  const deltas: number[] = [];
  for (let i = 1; i < closes.length; i++) {
    deltas.push(closes[i] - closes[i - 1]);
  }
  const recent = deltas.slice(-period);
  const avgGain = recent.reduce((s, d) => s + (d > 0 ? d : 0), 0) / period;
  const avgLoss = recent.reduce((s, d) => s + (d < 0 ? -d : 0), 0) / period;
  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export function computeBollinger(
  closes: number[],
  period = 20,
  stdDev = 2,
): [number, number, number] {
  if (closes.length < period) return [0, 0, 0];
  const recent = closes.slice(-period);
  const mid = recent.reduce((a, b) => a + b, 0) / period;
  const variance = recent.reduce((sum, x) => sum + (x - mid) ** 2, 0) / period;
  const std = Math.sqrt(variance);
  return [mid - stdDev * std, mid, mid + stdDev * std];
}

export function generateSignals(closes: number[], lastPrice: number): string[] {
  const signals: string[] = [];
  const sma20 = computeSma(closes, 20);
  const sma50 = computeSma(closes, 50);
  const rsi = computeRsi(closes);
  const [bbLower, , bbUpper] = computeBollinger(closes);

  if (sma20 && sma50) {
    signals.push(
      sma20 > sma50
        ? "MA Crossover: SMA20 > SMA50 → BULLISH"
        : "MA Crossover: SMA20 < SMA50 → BEARISH",
    );
  }

  if (rsi < 30) {
    signals.push(`RSI(${rsi.toFixed(1)}): Oversold → BUY signal`);
  } else if (rsi > 70) {
    signals.push(`RSI(${rsi.toFixed(1)}): Overbought → SELL signal`);
  } else {
    signals.push(`RSI(${rsi.toFixed(1)}): Neutral`);
  }

  if (bbLower && bbUpper) {
    if (lastPrice < bbLower) {
      signals.push(`Bollinger: Price below lower band (${bbLower.toFixed(2)}) → BUY signal`);
    } else if (lastPrice > bbUpper) {
      signals.push(`Bollinger: Price above upper band (${bbUpper.toFixed(2)}) → SELL signal`);
    } else {
      signals.push(`Bollinger: Price within bands (${bbLower.toFixed(2)} - ${bbUpper.toFixed(2)})`);
    }
  }

  return signals;
}
